### PR TITLE
Vmware nsxt for ocp4

### PR DESCRIPTION
**Pull request by mkelnermishal**

### CODE REVIEW

File: upi/vsphere/main.tf [hunk 1]
```diff
@@ -99,3 +110,26 @@ module "dns" {
   compute_count       = "${var.compute_count}"
   compute_ips         = ["${module.compute.ip_addresses}"]
 }
+
+module "nsx_network" {
+  source = "./nsx_network"
+
+  base_domain         = "${var.base_domain}"
```
Comment (salv-orlando) [inline upi/vsphere/main.tf]: indent does not seem right here

File: upi/vsphere/main.tf [hunk 2]
```diff
@@ -5,6 +5,17 @@ provider "vsphere" {
...
 
+provider "nsxt" {
+  host                     = "${var.nsx_manager}"
+  username                 = "${var.nsx_user}"
+  password                 = "${var.nsx_password}"
+  allow_unverified_ssl     = true
+  max_retries              = 10
+  retry_min_delay          = 500
+  retry_max_delay          = 5000
+  retry_on_status_codes    = [429]
```
Comment (salv-orlando) [inline upi/vsphere/main.tf]: It Might be worth retrying on 503 and 429

File: upi/vsphere/terraform.tfvars.example [hunk 3]
```diff
@@ -75,3 +78,20 @@ ipam_token = "TOKEN_FOR_THE_IPAM_SERVER"
 // The IP addresses to assign to the compute VMs. The length of this list must
 // match the value of compute_count.
 //compute_ips = ["10.0.0.30", "10.0.0.31", "10.0.0.32"]
+
+// NSX part for NSX-T network.
```
Comment (salv-orlando) [inline upi/vsphere/terraform.tfvars.example]: While many deployments will have NSX-T not all deployments will be on NSX-T. We need a variable to switch NSX-T on/off. If off, the NSX-T resources won't be created.

File: upi/vsphere/variables.tf [hunk 4]
```diff
@@ -131,3 +132,68 @@ variable "compute_ips" {
...
+  type = "string"
+}
+
+variable "logical_switch_ip_address" {
+  type = "string"
+}
+
+variable "transport_zone_id" {
+  type = "string"
+}
```
Comment (salv-orlando) [inline upi/vsphere/variables.tf]: add a newline at the end of the file please

File: upi/vsphere/variables.tf [hunk 5]
```diff
@@ -131,3 +132,68 @@ variable "compute_ips" {
...
+  type        = "string"
+  description = "NSX manager user for the environment."
+}
+
+variable "nsx_password" {
+  type        = "string"
+  description = "NSX manager password"
+}
+
+variable "ip_block_cidr" {
```
Comment (salv-orlando) [inline upi/vsphere/variables.tf]: all these variables deserve a description imho

File: upi/vsphere/variables.tf [hunk 6]
```diff
@@ -131,3 +132,68 @@ variable "compute_ips" {
...
+  type = "string"
+}
+
+variable "ip_pool_start" {
+  type = "string"
+}
+
+variable "ip_pool_end" {
```
Comment (salv-orlando) [inline upi/vsphere/variables.tf]: are these for the external ip pool? If so, prepend all of them with "external"
some of those are optional (like start and end) make sure this is made clear in the descriptin

File: upi/vsphere/nsx_network/variables.tf [hunk 7]
```diff
@@ -0,0 +1,77 @@
+variable "nsx_manager" {
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/variables.tf]: If I read the patch correctly these variables are also defined in upi/vsphere/variables.tf - is that intentional?

File: upi/vsphere/nsx_network/main.tf [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  admin_state       = "UP"
+  description       = "LP1 provisioned by Terraform"
+  display_name      = "LP1"
+  logical_switch_id = "${var.logical_switch}"
+}
+
+resource "nsxt_logical_router_downlink_port" "downlink_port" {
+  description                   = "DP1 provisioned by Terraform"
+  display_name                  = "DP1"
+  logical_router_id             = "${var.t1_router}"
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/main.tf]: it seems a bit weird that we don't create router and switch as well. What is the reason?

File: upi/vsphere/nsx_network/main.tf [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  linked_logical_switch_port_id = "${nsxt_logical_port.logical_port.id}"
+  ip_address                    = "${var.logical_switch_ip_address}"
+}
+
+
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/main.tf]: There should be a single blank line

File: upi/vsphere/nsx_network/main.tf [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+resource "nsxt_dhcp_server_profile" "dhcp_profile" {
+  description                 = "dhcp_profile provisioned by Terraform"
+  display_name                = "dhcp_profile"
+  edge_cluster_id             = "${var.nsx_edge_cluster}"
+}
+
+resource "nsxt_logical_dhcp_server" "logical_dhcp_server" {
+  display_name    = "logical_dhcp_server"
+  dhcp_profile_id = "${nsxt_dhcp_server_profile.dhcp_profile.id}"
+  dhcp_server_ip  = "${var.dhcp_server_ip}/24"
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/main.tf]: prefix should not be hardcoded. Can we have it as a variable?

File: upi/vsphere/nsx_network/main.tf [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  logical_switch_id = "${var.logical_switch}"
+  dhcp_server_id    = "${nsxt_logical_dhcp_server.logical_dhcp_server.id}"
+}
+
+resource "nsxt_dhcp_server_ip_pool" "dhcp_ip_pool" {
+  display_name = "ip pool"
+  description = "ip pool"
+  logical_dhcp_server_id = "${nsxt_logical_dhcp_server.logical_dhcp_server.id}"
+  gateway_ip = "${var.gateway_ip}"
+  lease_time = 180
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/main.tf]: I'm ok with hardcoding it, but why such a short lease?

File: upi/vsphere/nsx_network/main.tf [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  error_threshold = 98
+  warning_threshold = 70
+
+  ip_range {
+    start = "${var.ip_pool_start}"
+    end = "${var.ip_pool_end}"
+  }
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/main.tf]: I don't see where we configure default DHCP options. For OCP we will only need the gateway

File: upi/vsphere/nsx_network/main.tf [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  description  = "ip_block provisioned by Terraform"
+  display_name = "ip_block"
+  cidr         = "${var.ip_block_cidr}"
+}
+
+resource "nsxt_ip_block_subnet" "ip_block_subnet" {
+  description = "ip_block_subnet"
+  block_id    = "${nsxt_ip_block.ip_block.id}"
+  size        = 16
+}
```
Comment (salv-orlando) [inline upi/vsphere/nsx_network/main.tf]: I'm not sure why you're cutting a /16 subnet from the block

File: upi/vsphere/variables.tf (hunk 6 above)
Comment (mkelnermishal) [reply upi/vsphere/variables.tf]: these are for the IP pool defined for the DHCP, that will hold the cluster VMs

File: upi/vsphere/nsx_network/main.tf (hunk 8 above)
Comment (mkelnermishal) [reply upi/vsphere/nsx_network/main.tf]: Currently, the reason for that is the fact that the DNS that serves the installation and the cluster as to be in a network that is accessible to the VMs. therefore, the logical switch and t1 router need to be created in advance. This needs to be changed eventually.

File: upi/vsphere/nsx_network/main.tf (hunk 9 above)
Comment (mkelnermishal) [reply upi/vsphere/nsx_network/main.tf]: fixed.

File: upi/vsphere/nsx_network/main.tf (hunk 11 above)
Comment (mkelnermishal) [reply upi/vsphere/nsx_network/main.tf]: this was done for testing purposes, changing to a more reasonable value.

**Comment by abhinavdahiya** [comment]
Please open an GitHub issue that defines the problem so that there is a discussion.

**Comment by patrickdillon** [comment]
cc @jcpowermac just fyi

File: upi/vsphere/main.tf (hunk 1 above)
Comment (mkelnermishal) [reply upi/vsphere/main.tf]: fixed

File: upi/vsphere/main.tf (hunk 2 above)
Comment (mkelnermishal) [reply upi/vsphere/main.tf]: added

File: upi/vsphere/variables.tf (hunk 4 above)
Comment (mkelnermishal) [reply upi/vsphere/variables.tf]: added

File: upi/vsphere/variables.tf (hunk 5 above)
Comment (mkelnermishal) [reply upi/vsphere/variables.tf]: added

File: upi/vsphere/nsx_network/variables.tf (hunk 7 above)
Comment (mkelnermishal) [reply upi/vsphere/nsx_network/variables.tf]: removed. thanks

File: upi/vsphere/nsx_network/main.tf (hunk 8 above)
Comment (mkelnermishal) [reply upi/vsphere/nsx_network/main.tf]: added the proper elements up to t0 router.

**Pull request closed without merging**